beacon_node/store/src/hot_cold_store.rs

```diff
@@ -1794,34 +1895,46 @@ impl<E: EthSpec, Hot: ItemStore<E>, Cold: ItemStore<E>> HotColdDB<E, Hot, Cold>
     /// Returns `HDiffBuffer` for the specified slot, or `HDiffBuffer` for the `ReplayFrom` slot if
     /// the diff for the specified slot is not stored.
     fn load_hdiff_buffer_for_slot(&self, slot: Slot) -> Result<(Slot, HDiffBuffer), Error> {
-        if let Some(buffer) = self.hdiff_buffer_cache.lock().get(&slot) {
+        if let Some(buffer) = self.historic_state_cache.lock().get_hdiff_buffer(slot) {
             debug!(
                 self.log,
                 "Hit diff buffer cache";
                 "slot" => slot
             );
             metrics::inc_counter(&metrics::STORE_BEACON_HDIFF_BUFFER_CACHE_HIT);
             return Ok((slot, buffer.clone()));
-        } else {
-            metrics::inc_counter(&metrics::STORE_BEACON_HDIFF_BUFFER_CACHE_MISS);
         }
+        metrics::inc_counter(&metrics::STORE_BEACON_HDIFF_BUFFER_CACHE_MISS);
 
         // Load buffer for the previous state.
         // This amount of recursion (<10 levels) should be OK.
         let t = std::time::Instant::now();
         match self.hierarchy.storage_strategy(slot)? {
             // Base case.
             StorageStrategy::Snapshot => {
-                let state = self
+                let mut state = self
                     .load_cold_state_as_snapshot(slot)?
                     .ok_or(Error::MissingSnapshot(slot))?;
-                let buffer = HDiffBuffer::from_state(state);
+                let buffer = HDiffBuffer::from_state(state.clone());
 
-                self.hdiff_buffer_cache.lock().put(slot, buffer.clone());
+                let t = std::time::Instant::now();
+                state.build_all_caches(&self.spec)?;
                 debug!(
                     self.log,
-                    "Added diff buffer to cache";
-                    "load_time_ms" => t.elapsed().as_millis(),
+                    "Built caches for state";
+                    "target_slot" => slot,
+                    "build_time_ms" => t.elapsed().as_millis()
+                );
+
+                self.historic_state_cache
+                    .lock()
+                    .put_both(slot, state, buffer.clone());
+
+                let load_time_ms = t.elapsed().as_millis();
```

Comment:
This time is wrong currently because it doesn't include the hdiff buffer load time